feat: playing around with aw-sync on android and misc

0xbrayo opened the pull request:
<!-- ELLIPSIS_HIDDEN -->


> [!IMPORTANT]
> Enhances Android app with session-based event tracking, background sync services, and updated configurations for improved data accuracy and synchronization.
> 
>   - **Behavior**:
>     - Introduces session-based event tracking in `SessionEventWatcher.kt` and `UsageStatsWatcher.kt`, replacing heartbeat-based tracking for improved accuracy.
>     - Adds `BackgroundService.kt` to manage server and sync operations in the background.
>     - Implements `SyncScheduler.kt` and `SyncInterface.kt` for periodic data synchronization.
>     - Updates `MainActivity.kt` to start `BackgroundService` for continuous operation.
>   - **Configuration**:
>     - Updates `build.gradle` and `gradle.properties` to use Kotlin 1.9.0 and Gradle 8.14.
>     - Sets `minSdkVersion` to 26 and `targetSdkVersion` to 34 in `mobile/build.gradle`.
>     - Adds new permissions in `AndroidManifest.xml` for foreground services and data sync.
>   - **Testing**:
>     - Adds `WebWatcherTest.kt` and `ScreenshotTest.kt` for testing web activity tracking and UI.
>     - Introduces `CustomTabsWrapper.kt` for handling custom tab interactions in tests.
>   - **Misc**:
>     - Updates `README.md` with new build instructions.
>     - Adds `EventParsingWorker.kt` for scheduled event parsing using WorkManager.
>     - Removes `ChromeWatcher.kt` and replaces it with `WebWatcher.kt` for broader browser support.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=ActivityWatch%2Faw-android&utm_source=github&utm_medium=referral)<sup> for 38b89bfc10110d947cd4606f61580cf66ccdd21d. You can [customize](https://app.ellipsis.dev/ActivityWatch/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

0xbrayo committed:
docs: update deprecated magick command

0xbrayo committed:
docs: update deprecated magick command

0xbrayo committed:
chore: add 'sh' to README.md

0xbrayo committed:
chore: add 'sh' to README.md

0xbrayo committed:
chore: replace the default android.jks.age

0xbrayo committed:
chore: replace the default android.jks.age

0xbrayo committed:
refactor: refactored to improve accuracy of the timeline

0xbrayo committed:
refactor: refactored to improve accuracy of the timeline

konrad-krol-incubly committed:
feat: implemented generic web watcher with support of multiple web browsers

konrad-krol-incubly committed:
feat: implemented generic web watcher with support of multiple web browsers

konrad-krol-incubly committed:
build(deps): updated project dependencies, jvm target to 11, gradle to 8.14

konrad-krol-incubly committed:
build(deps): updated project dependencies, jvm target to 11, gradle to 8.14

0xbrayo committed:
feat:use device name as hostname

0xbrayo committed:
feat:use device name as hostname

0xbrayo committed:
chore: gradle version bump and misc

0xbrayo committed:
chore: gradle version bump and misc

0xbrayo committed:
feat(android): add SyncInterface.kt with JNI bindings for aw-sync

- Created SyncInterface class to interact with native aw-sync library
- Implemented JNI bindings for sync operations:
  * syncPullAll - pull from all hosts
  * syncPush - push local data
  * syncBoth - full bidirectional sync
  * getSyncDir - get sync directory path
- Async wrappers with callbacks for all sync operations
- Automatic sync directory setup in external storage
- Proper error handling and logging

Phase 4 implementation: Android app integration

0xbrayo committed:
feat(android): add SyncInterface.kt with JNI bindings for aw-sync

- Created SyncInterface class to interact with native aw-sync library
- Implemented JNI bindings for sync operations:
  * syncPullAll - pull from all hosts
  * syncPush - push local data
  * syncBoth - full bidirectional sync
  * getSyncDir - get sync directory path
- Async wrappers with callbacks for all sync operations
- Automatic sync directory setup in external storage
- Proper error handling and logging

Phase 4 implementation: Android app integration

0xbrayo committed:
feat(android): implement automatic background sync

- Modified SyncInterface to use Downloads/ActivityWatch/ directory
  * Static path for easy access by sync apps (Syncthing, etc.)
  * Falls back gracefully if storage not available

- Created SyncScheduler for automatic periodic syncing
  * First sync runs 1 minute after app starts
  * Subsequent syncs every 15 minutes while app is open
  * Runs syncBoth (pull + push) for full bidirectional sync
  * Logs all sync operations and results

- Integrated into MainActivity lifecycle
  * Starts scheduler in onCreate after server initialization
  * Stops scheduler in onDestroy to clean up resources

No UI needed - syncs automatically in background.
Users can access sync directory at: /sdcard/Download/ActivityWatch/

Phase 4 complete: Automatic background sync implementation

0xbrayo committed:
feat(android): implement automatic background sync

- Modified SyncInterface to use Downloads/ActivityWatch/ directory
  * Static path for easy access by sync apps (Syncthing, etc.)
  * Falls back gracefully if storage not available

- Created SyncScheduler for automatic periodic syncing
  * First sync runs 1 minute after app starts
  * Subsequent syncs every 15 minutes while app is open
  * Runs syncBoth (pull + push) for full bidirectional sync
  * Logs all sync operations and results

- Integrated into MainActivity lifecycle
  * Starts scheduler in onCreate after server initialization
  * Stops scheduler in onDestroy to clean up resources

No UI needed - syncs automatically in background.
Users can access sync directory at: /sdcard/Download/ActivityWatch/

Phase 4 complete: Automatic background sync implementation

0xbrayo committed:
feat: always active in the background

0xbrayo committed:
feat: always active in the background

0xbrayo committed:
chore: minimum api target bump

0xbrayo committed:
chore: minimum api target bump

0xbrayo commented:
@ellipsis-dev review

0xbrayo committed:
fix: fix launch on boot not working

0xbrayo committed:
fix: fix launch on boot not working

0xbrayo committed:
feat: experimental parsing of events in background

0xbrayo committed:
feat: experimental parsing of events in background

0xbrayo committed:
build: update makefile to package aw-sync.so

0xbrayo committed:
build: update makefile to package aw-sync.so

0xbrayo committed:
fix: fix single-instance errors on android

0xbrayo committed:
fix: fix single-instance errors on android

0xbrayo committed:
fix: fix doNotStrip property not working

0xbrayo committed:
fix: fix doNotStrip property not working

0xbrayo commented:
@ellipsis-dev review

0xbrayo commented:
@greptile-apps review

0xbrayo committed:
Update mobile/src/main/java/net/activitywatch/android/watcher/WebWatcher.kt

Co-authored-by: ellipsis-dev[bot] <65095814+ellipsis-dev[bot]@users.noreply.github.com>

0xbrayo committed:
Update mobile/src/main/java/net/activitywatch/android/watcher/WebWatcher.kt

Co-authored-by: ellipsis-dev[bot] <65095814+ellipsis-dev[bot]@users.noreply.github.com>

0xbrayo commented on mobile/src/main/java/net/activitywatch/android/watcher/SessionEventWatcher.kt:
```diff
@@ -0,0 +1,247 @@
+     */
+    fun sendSessionEvents() {
+        Log.w(TAG, "Starting SendSessionEventTask")
+        CoroutineScope(Dispatchers.IO).launch {
```
it just runs forever...

0xbrayo commented:
@greptileai review

0xbrayo commented:
@ErikBjare review, still not ready for merging but feedback would be appreciated. The PR is quite large unfortunately

0xbrayo committed:
style: fix linting warrnings

0xbrayo committed:
style: fix linting warrnings

harryob2 commented:
really looking forward to this PR. right now there is no way to get your android screen time data into your own database without building your own background service that polls UsageStats directly. RescueTime used to collect this data which you could then pull from their API, but they no longer support their Android app.

0xbrayo committed:
feat: add sync support methods to RustInterface

0xbrayo committed:
feat: add sync support methods to RustInterface

0xbrayo committed:
refactor: improve RustInterface initialization

0xbrayo committed:
refactor: improve RustInterface initialization

0xbrayo committed:
chore: update aw-server-rust submodule reference

0xbrayo committed:
chore: update aw-server-rust submodule reference

0xbrayo committed:
feat: implement category time widget

Initial implementation of home screen widget showing category time breakdown.

0xbrayo committed:
feat: implement category time widget

Initial implementation of home screen widget showing category time breakdown.

0xbrayo committed:
feat: enhance widget with visual indicators

Add bar backgrounds and progress dots for better visual feedback.

0xbrayo committed:
feat: enhance widget with visual indicators

Add bar backgrounds and progress dots for better visual feedback.

0xbrayo committed:
feat: implement widget background worker

Add CategoryTimeWidgetWorker to fetch and process data for widget display.

0xbrayo committed:
feat: implement widget background worker

Add CategoryTimeWidgetWorker to fetch and process data for widget display.

0xbrayo committed:
refactor: optimize widget layout for 4x1 size

0xbrayo committed:
refactor: optimize widget layout for 4x1 size

0xbrayo committed:
feat: add widget refresh animation

0xbrayo committed:
feat: add widget refresh animation

0xbrayo committed:
refactor: simplify widget styling and worker logic

Streamline the widget code and improve color scheme for better readability.

0xbrayo committed:
refactor: simplify widget styling and worker logic

Streamline the widget code and improve color scheme for better readability.

0xbrayo committed:
feat: media watcher

0xbrayo committed:
feat: media watcher

0xbrayo commented:
@TimeToBuildBob review this

TimeToBuildBob reviewed (commented):
Thanks for the tag! Took a look at this PR. Here's my review:

## Overall

This is a really exciting direction — getting `aw-sync` integrated via JNI so Android data can sync to a central AW instance would unlock a long-requested use case. The structure looks reasonable for an exploratory branch.

## Issues Found

### `applicationIdSuffix ".debug"` on release builds (likely bug)

In `mobile/build.gradle`:
```groovy
buildTypes {
    release {
+       applicationIdSuffix ".debug"
```

This would make the release APK install as `net.activitywatch.android.debug` instead of `net.activitywatch.android`, making it a different app — existing users couldn't upgrade. This looks like a dev/testing leftover that should be removed before this becomes mergeable.

### minSdkVersion raised from 24 → 26

This drops Android 7.0/7.1 (API 24/25) support. Worth confirming this is intentional — it's probably fine since those are very old (~2016), but it's a breaking change for existing users on those devices.

### CI failures

- **"Get latest versionCode"** — looks like a CI infrastructure issue unrelated to the code changes
- **"Test E2E (Pixel_API_27_AOSP)"** — worth investigating; API 27 is above the new minSdkVersion of 26 so it should work in theory

## What Looks Good

- The `SyncInterface` structure is clean — sets up XDG env vars, loads `aw_sync` JNI lib, exposes async wrappers
- `SyncScheduler` dispatches actual sync to background thread via `Executors.newSingleThreadExecutor()`, so the main looper usage is just for scheduling (not blocking)
- Makefile rules for `libaw_sync.so` follow the same pattern as `libaw_server.so`
- Bumping Gradle to 8.14 and Kotlin to 1.9.0 is a good maintenance improvement

## Suggestions

1. Remove `applicationIdSuffix ".debug"` from release builds
2. Consider using `WorkManager` for the periodic sync scheduling instead of `Handler.postDelayed` — it's more resilient to app restarts and doze mode
3. The hardcoded 15-minute interval in `SyncScheduler` should probably be configurable or at least a named constant

Looking forward to seeing this evolve!

0xbrayo commented:
@ellipsis-dev review

0xbrayo commented:
@TimeToBuildBob review this PR. suggest any changes that might be necessary.

ErikBjare commented on mobile/build.gradle:
```diff
@@ -42,20 +43,29 @@ android {
     }
+
     namespace 'net.activitywatch.android'
     // Never got this to work...
```
Works now?

0xbrayo replied:
yes it works now.

ErikBjare commented on mobile/src/main/java/net/activitywatch/android/RustInterface.kt:
```diff
@@ -82,25 +86,74 @@ class RustInterface constructor(context: Context? = null) {
+    /**
+     * Insert a discrete event that will not be merged with other events.
+     *
+     * This method is preferred for accurate app usage tracking because:
```
This makes a lot of sense, idk why I was trying to turn range-based events into heartbeats.

ErikBjare commented on mobile/src/main/java/net/activitywatch/android/SyncInterface.kt:
```diff
@@ -0,0 +1,105 @@
+        callback: (Boolean, String) -> Unit,
+        syncFn: () -> String
+    ) {
+        val executor = Executors.newSingleThreadExecutor()
```
I guess they aren't?

ErikBjare commented on mobile/src/main/java/net/activitywatch/android/widget/CategoryTimeWidgetWorker.kt:
```diff
@@ -0,0 +1,340 @@
+)
+
+/**
+ * Worker that updates the widget in the background every 30 minutes
```
Any way to do this more slightly intelligently to better do the whole recency vs low-resource usage tradeoff? 30min can be quite a long delay.

0xbrayo replied:
There is a limitation set by android for how frequently widgets can update. 15 mins is the lowest we can get to, using a workmanager (as used in this case). 30 mins is the minimum for android system updates (the update is triggered by android itself, doesn't require the app to be running in the background). I initially started with the latter and I guess kept the time constraint. I can update it to 15 mins :)

ErikBjare reviewed (commented):
Very impressive!

From a initial skim it looks great overall! Should be a major improvement.

I would like to see CI pass next, maybe I and/or Bob can help with that.

0xbrayo committed:
chore: update aw-server-rust submodule to upstream main to fix aw-webui submodule fetch in CI

0xbrayo committed:
feat: bump ci versions

0xbrayo committed:
feat: bump ci versions

0xbrayo committed:
fix(ci): update sdkmanager, avdmanager, and emulator paths for setup-android@v3

0xbrayo committed:
fix(ci): update sdkmanager, avdmanager, and emulator paths for setup-android@v3

0xbrayo committed:
fix(ci): remove fail-on-cache-miss from cache restore

0xbrayo committed:
fix(ci): remove fail-on-cache-miss from cache restore

0xbrayo committed:
fix(ci): restore missing path key and set fail-on-cache-miss: false in cache restore steps

0xbrayo committed:
fix(ci): restore missing path key and set fail-on-cache-miss: false in cache restore steps

0xbrayo committed:
refactor(ci): replace manual emulator setup with reactivecircus/android-emulator-runner@v2

refactor(ci): simplify test-e2e with api-level 29, google_apis, screenshot in script

0xbrayo committed:
refactor(ci): replace manual emulator setup with reactivecircus/android-emulator-runner@v2

refactor(ci): simplify test-e2e with api-level 29, google_apis, screenshot in script

0xbrayo committed:
fix(ci): launch app, sleep 1s, screenshot and logcat in emulator-runner script

0xbrayo committed:
fix(ci): launch app, sleep 1s, screenshot and logcat in emulator-runner script

0xbrayo committed:
fix(ci): correct am start intent to use debug package + full class name

0xbrayo committed:
fix(ci): correct am start intent to use debug package + full class name

0xbrayo committed:
fix(ci): install apk explicitly and use monkey to launch app before screenshot

0xbrayo committed:
fix(ci): install apk explicitly and use monkey to launch app before screenshot

0xbrayo committed:
fix(ci): bypass onboarding, grant usage stats perm, unlock screen before screenshot

0xbrayo committed:
fix(ci): bypass onboarding, grant usage stats perm, unlock screen before screenshot

0xbrayo committed:
fix(ci): pass jniLibs via artifact upload/download instead of cache

0xbrayo committed:
fix(ci): pass jniLibs via artifact upload/download instead of cache

0xbrayo commented:
@TimeToBuildBob the screenshots in test E2E are blank, diagnose and suggest a fix.

0xbrayo committed:
chore: update settings.gradle

0xbrayo committed:
chore: update settings.gradle

0xbrayo committed:
feat: update widget every 15 mins

0xbrayo committed:
feat: update widget every 15 mins

TimeToBuildBob commented:
## Screenshot Diagnosis

The blank screenshots aren't a capture bug — they're blank because the app crashes before anything renders.

### Root cause: `RemoteServiceException`

```
android.app.RemoteServiceException: Context.startForegroundService() did not then call 
Service.startForeground(): ServiceRecord{...net.activitywatch.android.BackgroundService}
```

`MainActivity.onCreate()` calls `startForegroundService(BackgroundService)` on line 61. Android gives the service a ~5-second window to call `startForeground()`, but the `BackgroundService` is crashing before that happens.

### Likely culprit: `RustInterface(this)` in `BackgroundService.onCreate()`

Looking at `onCreate()`:
```kotlin
override fun onCreate() {
    super.onCreate()
    rustInterface = RustInterface(this)  // ← This likely fails in test
    syncScheduler = SyncScheduler(this)
}
```

If `RustInterface(this)` throws or hangs (native library not initialized in test context, missing JNI libs on emulator, etc.), `onCreate()` never completes, so `onStartCommand()` is never called, so `startForeground()` is never called → `RemoteServiceException`.

### The screenshot timing confirms this

The test does:
1. Launch `MainActivity` → triggers `startForegroundService(BackgroundService)` → crash
2. `Thread.sleep(5000)` → waits while the app is dead
3. `takeScreenshot()` → captures a blank/crashed screen (1183 bytes = mostly empty PNG header)

### Suggested fixes

**Option A** (quick): Move `startForeground()` to `onCreate()` before `RustInterface()`:
```kotlin
override fun onCreate() {
    super.onCreate()
    createNotificationChannel()
    startForeground(NOTIFICATION_ID, createNotification())  // ← First!
    rustInterface = RustInterface(this)
    syncScheduler = SyncScheduler(this)
}
```
This ensures the foreground notification is posted before any potentially-failing init.

**Option B** (better): Add a try-catch around `RustInterface` init in `onCreate()`, and/or add `FOREGROUND_SERVICE` + `FOREGROUND_SERVICE_DATA_SYNC` to the test's `GrantPermissionRule`.

**For the CI failure** (`Get latest versionCode` → `Load Fastlane secrets`): That's a secrets management issue unrelated to the code — the Fastlane age key isn't configured for this branch in CI.

0xbrayo commented:
@TimeToBuildBob a better fix would be to ensure the jnis are always present in the emulator before compiling and launching the app.

0xbrayo committed:
feat: update widget every 5 mins

0xbrayo committed:
feat: update widget every 5 mins

0xbrayo commented:
@greptile review

0xbrayo committed:
feat: unlock watcher

0xbrayo committed:
feat: unlock watcher

0xbrayo committed:
chore: linting diff

0xbrayo committed:
chore: linting diff

0xbrayo commented:
After some testing on my phone, seems exact alarms are really unreliable, android eventually stops scheduling them due to the frequency (every 5 mins), a workaround could be to allow unrestricted battery usage, but I don't think it's worth tbh. I tested on a Samsung phone, which is known to quite aggressive about killing background processes. On the emulator it works no issues, though very few processes are running on there anyway.

0xbrayo committed:
feat: migrate legacy db hostnames

0xbrayo committed:
feat: migrate legacy db hostnames

0xbrayo commented:
depends on https://github.com/ActivityWatch/aw-server-rust/pull/577

0xbrayo commented:
Tested the migration on the emulator, worked seamlessly for me :)

TimeToBuildBob commented:
Followed up on the failing E2E job. The blank screenshot theory was a red herring — this failure is happening *before the app launches*.

## Root cause
The install step is failing with:

```txt
INSTALL_FAILED_NO_MATCHING_ABIS: Failed to extract native libraries, res=-113
```

From the same job log:

```txt
mobile/src/main/jniLibs/
├── arm64-v8a/
│   ├── libaw_server.so
│   └── libaw_sync.so
└── armeabi-v7a/
    └── libaw_server.so
```

So the artifact downloaded into the E2E job only contains ARM JNI libs. The emulator job is running an x86_64 emulator, so `adb install` rejects the APK before `monkey` or the screenshot step ever gets a chance.

## What to fix
The CI needs to make sure the E2E job has x86_64 JNI libs available when building/installing the debug APK.

Concretely, one of these needs to happen:

1. **Preferred:** make the `jniLibs` artifact include `x86/` and `x86_64/` variants (`libaw_server.so` and `libaw_sync.so`) before the E2E job downloads it.
2. Or switch the E2E emulator to an ARM system image that matches the uploaded libs.
3. Or rebuild/copy the missing x86_64 JNI libs inside the E2E job before `make test-e2e`.

## Extra signal
There’s also a suspicious strip warning earlier in the same build:

```txt
Unable to strip ... libaw_server.so
llvm-strip: error: ... armeabi-v7a/libaw_server.so: section header table goes past the end of the file
```

That might be a separate issue with how the ARM `.so` is produced or packaged, but it’s **not** what’s blocking this E2E job right now. The hard blocker is the ABI mismatch.

So: the right fix isn’t "make screenshots work" — it’s "ship x86_64 JNI libs into the emulator job."

TimeToBuildBob committed:
ci: skip secrets-dependent steps on fork PRs

Fork PRs cannot access repository secrets (KEY_FASTLANE_API,
KEY_ANDROID_JKS), which caused get-versionCode to always fail on
external contributions.

Fix:
- Add `if` conditions to Ruby/fastlane/age/update_version steps so they
  only run on the main repo or tag/push triggers.
- Move `Output versionCode` to always run, reading directly from
  build.gradle as a fallback when fastlane is skipped.
- Skip build-apk entirely on fork PRs since signing keys are unavailable.

Result: fork PRs now pass build-rust, test, and test-e2e without
hitting the secrets wall. The versionCode/APK lane is silently skipped
rather than failing loudly with credential errors.

Addresses: https://github.com/ActivityWatch/aw-android/pull/139#issuecomment-4512051883

TimeToBuildBob commented:
Opened ActivityWatch/aw-android#159 with the CI fix.

The root cause: the `get-versionCode` job requires `KEY_FASTLANE_API` to decrypt the Google Play API key, but that secret isn't available to fork PRs (GitHub's security policy). The fix adds `if: github.event_name != 'pull_request'` to all the Fastlane setup/secrets/update steps. For PRs the job falls through to reading `versionCode` directly from `mobile/build.gradle` — the same approach `build-only.yml` already uses.

Once #159 merges, CI should pass for this PR and others from forks.

TimeToBuildBob committed:
fix(sync): handle missing aw-sync library

TimeToBuildBob commented on mobile/src/main/java/net/activitywatch/android/SyncInterface.kt:
```diff
@@ -0,0 +1,115 @@
+import java.io.File
+import java.util.concurrent.Executors
+
+private const val TAG = "SyncInterface"
+
+class SyncInterface(context: Context) {
+    private val appContext: Context = context.applicationContext
+    private val syncDir: String
+    
+    init {
+        // Use Downloads folder for easy user access: /sdcard/Download/ActivityWatch/
+        val downloadsDir = android.os.Environment.getExternalStoragePublicDirectory(
+            android.os.Environment.DIRECTORY_DOWNLOADS
+        )
+        syncDir = "$downloadsDir/ActivityWatch"
+        Os.setenv("AW_SYNC_DIR", syncDir, true)
+        
+        // Set XDG environment variables to app-writable paths
+        // This is required for aw-client-rust (used by aw-sync) to create lock files
+        val cacheDir = context.cacheDir.absolutePath
+        val filesDir = context.filesDir.absolutePath
+        
+        Os.setenv("XDG_CACHE_HOME", cacheDir, true)
+        Os.setenv("XDG_CONFIG_HOME", "$filesDir/config", true)
+        Os.setenv("XDG_DATA_HOME", "$filesDir/data", true)
+        
+        // Create sync directory if it doesn't exist
+        File(syncDir).mkdirs()
+        
+        System.loadLibrary("aw_sync")
+        Log.i(TAG, "aw-sync initialized with sync dir: $syncDir")
+    }
+    
+    // Native JNI functions
+    private external fun syncPullAll(port: Int, hostname: String): String
```
Fixed in 820f5c5: AW_SYNC_DIR now resolves to an app-writable external-files sync directory with an internal-files fallback instead of public Downloads.

TimeToBuildBob commented on mobile/src/main/java/net/activitywatch/android/SyncInterface.kt:
```diff
@@ -0,0 +1,115 @@
+        File(syncDir).mkdirs()
+        
+        System.loadLibrary("aw_sync")
+        Log.i(TAG, "aw-sync initialized with sync dir: $syncDir")
```
Fixed in 820f5c5: SyncScheduler and SyncAlarmReceiver now catch UnsatisfiedLinkError and disable/skip sync instead of crashing the app when libaw_sync is missing.

TimeToBuildBob committed:
fix(ci): stop masking e2e failures

TimeToBuildBob committed:
fix(sync): avoid replaying last session event

TimeToBuildBob commented on mobile/src/main/java/net/activitywatch/android/watcher/SessionEventWatcher.kt:
```diff
@@ -0,0 +1,276 @@
+        Log.w(TAG, "lastUpdated: ${lastUpdated?.toString() ?: "never"}")
+
+        val startTimestamp = lastUpdated?.toEpochMilli() ?: 0L
+        val sessions = sessionParser.parseUsageEventsSince(startTimestamp)
```
Fixed in 8ff553b.

The issue was real: `processEventsSinceLastUpdate()` was reusing the exact start timestamp of the last stored session, and `UsageStatsManager.queryEvents(start, end)` treats `start` as inclusive. That caused the same `ACTIVITY_RESUMED` event to be reparsed and reinserted on the next run.

I changed the incremental cursor to start at `lastUpdated + 1ms` and reused that same adjusted cursor in `getSessionsSinceLastUpdate()`, so we stop replaying the last stored session while still picking up later events normally.

0xbrayo commented on mobile/build.gradle:
```diff
@@ -29,11 +29,13 @@ android {
             ndk {
                 debugSymbolLevel 'SYMBOL_TABLE'
             }
+            signingConfig signingConfigs.debug
```
artifacts from my experiments,

TimeToBuildBob replied:
Fixed in 4d9c2cd. Removed `signingConfig signingConfigs.debug` and the blank `applicationIdSuffix` from the release build type — neither was in master, both were added locally during experimentation. Release APKs for distribution go through Fastlane with the proper keystore anyway.

TimeToBuildBob replied:
False positive. The current `mobile/build.gradle` does not add `signingConfig signingConfigs.debug` to the release build type — the PR diff only touches `minSdkVersion`, Java version, lint baseline, and dependencies. The release block is unchanged from master.

TimeToBuildBob replied:
Correction: this was real at commit 8ff553b (where it was introduced), not a false positive. It's been removed from the release block by the current HEAD (4d9c2cd) — the release buildType only has minifyEnabled, proguardFiles, and ndk/debugSymbolLevel. No further action needed.

TimeToBuildBob replied:
Fixed — `signingConfig signingConfigs.debug` is no longer present in the release build type. The release block now only has `minifyEnabled false`, `proguardFiles`, and `ndk { debugSymbolLevel 'SYMBOL_TABLE' }`. Release APKs will get Gradle's default unsigned output (as intended for F-Droid and Play Store CI pipelines).

TimeToBuildBob committed:
fix(ci): preinstall ndk before emulator runner

TimeToBuildBob committed:
fix: address two greptile P1 findings

- remove signingConfig signingConfigs.debug from release build type
  (debug-signed APKs are rejected by Play Store and F-Droid)
- remove blank applicationIdSuffix from release block
- fix widget refresh showing stale data: sendHeartbeats() fires a
  fire-and-forget coroutine so the widget update was racing against
  event insertion; add sendSessionEventsSuspend()/sendHeartbeatsSuspend()
  and restructure ACTION_REFRESH to use goAsync() + await before
  calling updateWidgetWithRefreshButton

TimeToBuildBob commented on mobile/src/main/java/net/activitywatch/android/widget/CategoryTimeWidgetProvider.kt:
```diff
@@ -0,0 +1,173 @@
+                
+                val appWidgetManager = AppWidgetManager.getInstance(context)
+                val componentName = ComponentName(context, CategoryTimeWidgetProvider::class.java)
+                val appWidgetIds = appWidgetManager.getAppWidgetIds(componentName)
+                
+                // Show loading indicator first
+                for (appWidgetId in appWidgetIds) {
+                    showLoadingState(context, appWidgetManager, appWidgetId)
+                }
+                
+                // Re-parse usage events
+                try {
+                    val usageStatsWatcher = UsageStatsWatcher(context)
+                    usageStatsWatcher.sendHeartbeats()
+                    Log.d(TAG, "Triggered usage events re-parsing")
+                } catch (e: Exception) {
+                    Log.e(TAG, "Error re-parsing usage events", e)
+                }
+                
+                // Then update with fresh data
+                for (appWidgetId in appWidgetIds) {
+                    updateWidgetWithRefreshButton(context, appWidgetManager, appWidgetId)
+                }
+            }
```
Fixed in 4d9c2cd. Added `sendSessionEventsSuspend()` to `SessionEventWatcher` (suspend wrapper around the existing sync `processEventsSinceLastUpdate()`) and `sendHeartbeatsSuspend()` to `UsageStatsWatcher`, then restructured the `ACTION_REFRESH` handler to use `goAsync()` so the BroadcastReceiver lifetime extends past `onReceive()` and the widget update runs after the coroutine completes.

TimeToBuildBob replied:
Fixed in 820f5c5 and confirmed in current HEAD. The refresh path now calls `sendHeartbeatsSuspend()` inside a `goAsync()` coroutine — widget update only runs after the coroutine completes, so the data is persisted before the query.

TimeToBuildBob replied:
Fixed — the Greptile 4/5 summary (latest review on current commit c4776a1c) confirms: 'the widget refresh path correctly awaits persistence via `sendHeartbeatsSuspend()` with `goAsync()`'.

TimeToBuildBob committed:
fix(ci): use default emulator target to avoid google_apis flakiness

google_apis on API 29 triggers ShellCommandUnresponsiveException on the
ubicloud runner — the heavier system image causes ADB to become
unresponsive during connectedDebugAndroidTest. The default target has no
Google Play Services but is significantly more stable on CI.

ErikBjare commented on mobile/src/main/java/net/activitywatch/android/watcher/SessionEventWatcher.kt:
```diff
@@ -0,0 +1,288 @@
+
+        var eventsSent = 0
+        for (session in sessions) {
+            insertSessionAsEvent(session)
```
Maybe we should finally do this and rip the bandaid off. The -test suffix was never meant to make it into prod.

0xbrayo committed:
fix(ci): skip deprecated 'tools' package in setup-android@v3

The 'tools' (Android SDK Tools) package was removed/corrupted upstream
and now fails with 'Error on ZipFile unknown archive'. Only install
platform-tools which is what we actually need.

0xbrayo committed:
fix: rename production bucket from aw-watcher-android-test to aw-watcher-android

The test suffix caused all real user app-usage data to be stored under
a bucket name that queries and dashboards targeting aw-watcher-android
would never find.

0xbrayo committed:
fix: use KEEP policy for EventParsingWorker, remove dead widget code

BackgroundService: ExistingPeriodicWorkPolicy.UPDATE was resetting the
nightly EventParsingWorker schedule on every app open. If the app was
used after midnight the job would slip 24 hours. Changed to KEEP so the
first-scheduled midnight alarm is never disturbed by subsequent starts.

CategoryTimeWidgetWorker: the entire companion object (schedulePeriodicUpdates,
cancelPeriodicUpdates, updateAllWidgets, updateSingleWidget, and all chart/
parsing helpers) was unreachable — the Provider uses AlarmManager and
CategoryTimeWidgetUpdater for all updates. Removed the dead 300-line
block; doWork() now delegates to CategoryTimeWidgetUpdater directly.

0xbrayo committed:
fix: address greptile review feedback (greploop iteration 1)

- MainActivity: remove unused lateinit syncScheduler (BackgroundService
  owns SyncScheduler now) and dead test comment
- CategoryTimeWidgetProvider: move widget update loop inside try block
  and guard pendingResult.finish() with finally so it always fires
- CategoryTimeWidgetProvider: make ACTION_REFRESH internal so Updater
  can reference it without duplicating the string literal
- CategoryTimeWidgetUpdater: replace hardcoded action string with the
  shared ACTION_REFRESH constant
- SessionEventWatcher: replace misleading W-level log on empty bucket
  (normal on first run) with a silent null return; keep the warning only
  for the truly unexpected case of >1 events returned

0xbrayo committed:
fix: dispatch onUpdate off main thread to avoid ANR (greploop iteration 2)

onUpdate called updateSingleWidget synchronously, which calls
ri.androidQuery() — a JNI call that makes a local HTTP request.
Wrap in goAsync() + IO coroutine, consistent with ACTION_REFRESH
and ACTION_PERIODIC_UPDATE handlers in the same class.

0xbrayo committed:
fix: prevent duplicate session insertion from concurrent callers (greploop iteration 3)

onResume and EventParsingWorker each create a new SessionEventWatcher,
read the same cursor from the server, and insert the same sessions.
Added a companion-object ReentrantLock with tryLock so a concurrent
call skips immediately rather than queuing behind and then re-inserting
the same window of sessions.

0xbrayo committed:
fix: pass foreground service type to startForeground on Android 14+ (greploop iteration 4)

Android 14 (API 34) requires the foreground service type to be declared
in both the manifest (already done: foregroundServiceType=dataSync) and
the startForeground() runtime call. The 2-arg call throws on API 34+.
Use ServiceCompat.startForeground with FOREGROUND_SERVICE_TYPE_DATA_SYNC
for API 29+ and fall back to 0 on older versions.

0xbrayo committed:
fix: register SyncAlarmReceiver in manifest (greploop iteration 5)

SyncAlarmReceiver handled BOOT_COMPLETED and SYNC_ALARM but was never
declared in AndroidManifest.xml, so Android never delivered either
broadcast. Boot-persistence and alarm-triggered sync were silently
non-functional on all devices.

0xbrayo committed:
fix: prevent async work from being orphaned in SyncAlarmReceiver (greploop final)

SYNC_ALARM: call goAsync() and finish pendingResult in the sync callback
(and in catch blocks) so the process is not killed before the executor
thread completes.

BOOT_COMPLETED: replace SyncScheduler(context).start() — which posts a
60s Handler.postDelayed that dies with the process — with
startForegroundService(BackgroundService). BackgroundService already owns
the SyncScheduler lifecycle and keeps the process alive as a foreground
service.

0xbrayo committed:
fix: bind session-event coroutine to Activity lifecycle; remove dual BOOT_COMPLETED (greploop final+1)

- Delete sendSessionEvents() (unmanaged CoroutineScope) from SessionEventWatcher
- Switch MainActivity.onResume() from sendHeartbeats() to
  lifecycleScope.launch { sendHeartbeatsSuspend() }, binding the
  coroutine to the Activity's lifecycle so Android can cancel it cleanly
- Remove BOOT_COMPLETED from SyncAlarmReceiver's intent-filter; AlarmReceiver
  already starts BackgroundService on boot, preventing a duplicate start

0xbrayo committed:
fix: restore background collection for session mode via goAsync in AlarmReceiver

AlarmReceiver.LOG_DATA path now calls sendHeartbeatsSuspend() inside
goAsync() + CoroutineScope(Dispatchers.IO) so the hourly alarm-driven
background collection works correctly in session mode.

sendHeartbeats() is simplified to the legacy (non-session) path only,
since AlarmReceiver now uses the suspend API directly.

0xbrayo committed:
fix: prevent concurrent syncs and unconditional hostname migration mark

SyncScheduler: move next-sync scheduling inside the syncBothAsync callback
so the 15-minute timer starts only after the previous sync completes,
preventing concurrent JNI calls to the native sync library.

BackgroundService: only call setHostnameMigrated() when migrateHostname()
does not return an error string, so a transient failure (server not yet
ready) does not permanently suppress the migration.

0xbrayo committed:
fix: wire AlarmManager fallback in SyncScheduler so SYNC_ALARM is reachable

SyncScheduler.start() now also schedules a 15-minute inexact repeating
AlarmManager alarm that fires SyncAlarmReceiver.SYNC_ALARM. This provides
a sync fallback when BackgroundService is killed by the OS; the Handler
chain remains the primary path while the service is alive.

SyncScheduler.stop() cancels the alarm to avoid sync after explicit stop.

0xbrayo committed:
fix: run hostname migration off main thread to avoid ANR; fix midnight calc

migrateHostname() is a blocking JNI call that can connect to the local
aw-server. Running it on the main thread in onStartCommand risks ANR when
the server is already running. Move it to Dispatchers.IO via a coroutine.

Also fix scheduleEventParsing(): reset Calendar.MILLISECOND to 0 so the
initial delay to midnight is exact rather than off by up to 999ms.

0xbrayo committed:
fix: restore widget periodic alarm in onUpdate (covers post-reboot case)

Android calls onUpdate for existing widget instances after device reboot.
Without calling schedulePeriodicUpdates() there, the 5-minute AlarmManager
chain is never re-created and background widget refreshes stop permanently
until the user removes and re-adds the widget.

0xbrayo committed:
fix: recycle AccessibilityNodeInfo in WebWatcher.findWebView (API 26-32 leak)

Nodes obtained via getChild() must be recycled when no longer needed on
API 26-32 (on 33+ recycle() is a no-op). Replace the lazy mapNotNull
chain (which acquired nodes and never recycled null-returning branches)
with an explicit loop that:
  - recycles child nodes whose subtrees contain no WebView
  - recycles intermediate ancestors when the returned node is deeper

0xbrayo committed:
fix: move MediaWatcher polling and JNI heartbeat off main thread

Replace Handler(Looper.getMainLooper()) with a dedicated HandlerThread.
heartbeatHelper() is a blocking JNI call; running it on the main looper
risks ANR during active media playback. The HandlerThread is quit safely
in onDestroy() to avoid thread leaks.

0xbrayo committed:
fix: prevent concurrent syncBoth() JNI calls via shared AtomicBoolean

SyncScheduler.start() arms both a Handler chain (15 min) and an AlarmManager
fallback (also 15 min). A slow initial sync can cause both to fire simultaneously,
calling syncBoth() on the same native library and sync directory from two threads.

Add a companion-object AtomicBoolean 'syncInFlight' shared across all
SyncInterface instances. syncBothAsync() uses compareAndSet to acquire it and
clears it in the callback, so any concurrent entry — whether from the Handler
chain, SyncAlarmReceiver, or a future caller — is skipped with a log.

0xbrayo committed:
fix: parse migrateHostname result as JSON instead of string-contains

Replace the fragile contains("error") check with a proper JSON parse of
the native response. If the JSON has "success": false, retry on next
start. If the response is not valid JSON, treat it as success (native
lib returned without signalling failure). This prevents a native response
that incidentally contains the word "error" from permanently blocking
setHostnameMigrated().

0xbrayo committed:
fix: pin all MediaWatcher callbacks to handlerThread to prevent data races

registerCallback() and addOnActiveSessionsChangedListener() are now passed
the handlerThread handler so all mutations of activeControllers,
activeCallbacks, and lastEventKeys happen on a single thread — the same
HandlerThread that runs the polling loop. This eliminates the
ConcurrentModificationException risk introduced when the polling was moved
off the main thread.

0xbrayo committed:
fix: add lifecycle-runtime-ktx dep, move SyncInterface init off main thread, deprecate sendHeartbeats()

build.gradle: add lifecycle-runtime-ktx:2.8.7 as an explicit dependency
  (required for lifecycleScope extension used in MainActivity.onResume).

SyncAlarmReceiver: move SyncInterface creation inside CoroutineScope(IO) so
  the file-system operations in SyncInterface.init{} don't run on the main
  thread; keeps the goAsync()+pendingResult.finish() lifecycle guarantee.

UsageStatsWatcher.sendHeartbeats(): mark @Deprecated with a ReplaceWith hint
  pointing to sendHeartbeatsSuspend(); the function is a no-op in session
  mode and callers should migrate to the suspend variant.

0xbrayo committed:
fix: keep AlarmManager fallback alarm alive when BackgroundService stops

SyncScheduler.stop() was cancelling the AlarmManager fallback, which defeats
its purpose on OEM devices that suppress START_STICKY restarts: if onDestroy()
fires before the process dies, the alarm is gone and sync never recovers.

Remove cancelAlarm() from stop() and delete the now-unused helper. The alarm
is registered once on start() and lives until Android cancels it on app
uninstall, ensuring SyncAlarmReceiver can always fire as a fallback.

0xbrayo committed:
fix: guard EventParsingWorker against epoch-0 cursor reset on cold server

If the aw-server is not running when the nightly WorkManager job fires,
RustInterface.getEventsJSON() returns an empty JSONArray (silently swallowing
the error), causing getLastEventTime() to return null, the cursor to reset
to epoch 0, and up to 30 days of UsageStats sessions to be re-inserted.

Add a server-availability check via getBucketsJSON() before calling
processEventsSinceLastUpdate(). An empty bucket list or JSONException both
indicate the server isn't ready; return Result.retry() so WorkManager
reschedules the job rather than corrupting the bucket with duplicates.

0xbrayo committed:
fix: remove test() debug method, clarify insertEvent KDoc, document suspend no-op

RustInterface.test(): remove the debug method with hardcoded 2026-01-31
timestamps; it created a 'test' bucket on any call and was never called
from production code.

insertEvent KDoc: rewrite to accurately describe the implementation
(heartbeat with pulsetime=0) rather than implying a separate server endpoint.

sendHeartbeatsSuspend(): add KDoc explaining the intentional legacy-mode
no-op so callers on the non-session path know to use sendHeartbeats() instead.

0xbrayo committed:
fix: use conservative defaults for hostname migration success detection

optBoolean("success", true) and catch { true } both silently treat
non-JSON or key-absent responses as successful migration, permanently
blocking retries if the server wasn't ready on first launch.

Change both to false so that:
- JSON without a "success" key → retry next start
- Non-JSON response (server error/not ready) → retry next start
- JSON with "success": true → mark migrated (only explicit success counts)

0xbrayo committed:
fix: distinguish server-down from fresh-install in EventParsingWorker guard

The previous guard returned Result.retry() on empty getBucketsJSON(), which
on a fresh install (no buckets yet) meant the worker never progressed and the
bucket was never created — background collection was permanently broken until
the user opened the app.

Fix: only retry on JSONException (non-JSON response = server not reachable).
An empty JSON object '{}' means the server IS up with no buckets yet — proceed
and let processEventsSinceLastUpdate() create the aw-watcher-android bucket.

0xbrayo committed:
fix: remove dead CategoryTimeWidgetWorker and move SyncInterface init off main thread

CategoryTimeWidgetWorker: dead code, never enqueued. Widget updates are driven
by the self-rescheduling AlarmManager alarm in CategoryTimeWidgetProvider.

SyncScheduler.start(): wrap SyncInterface construction in Dispatchers.IO to
avoid blocking the main thread with JNI calls (getDeviceName) and file I/O
(XDG env var resolution, sync dir lookup) during cold starts.

0xbrayo committed:
ci: fail hard on jniLibs cache miss instead of silently skipping tests

Previously, cache-hit == 'true' guards meant a jniLibs cache miss caused
the 'Check that jniLibs present' and 'Test' steps to be silently skipped
while the job reported success. This hid missing test coverage on PRs that
bump the aw-server-rust submodule.

Fix: fail-on-cache-miss: true causes the cache/restore step itself to fail
on miss, so CI fails visibly rather than silently skipping unit tests.

0xbrayo committed:
fix: remove bulk-import helpers with no deduplication guard

sendSessionEventsForDay, sendSessionEventsForPeriod, sendSessionEventsForLastDays,
forceRefreshToday, insertSessionsAsEvents, getTimelineForDay, and
getSessionsSinceLastUpdate had no callers outside their own files and would
produce duplicate events if invoked more than once for the same date range.

The production data path is processEventsSinceLastUpdate() which uses a
persistent cursor and ReentrantLock.tryLock() to prevent duplicates — the
removed helpers bypassed both guards.

0xbrayo committed:
fix: remove getTodayTimeline() that called deleted getTimelineForDay()

The previous commit removed getTimelineForDay() from SessionEventWatcher
but left getTodayTimeline() in UsageStatsWatcher calling it, causing a
compilation error. getTodayTimeline() itself has no callers, so remove it.

0xbrayo committed:
fix: use ConcurrentHashMap in MediaWatcher to prevent map-modification race

onDestroy() runs on the main thread and clears activeControllers/activeCallbacks
while pollActiveSessions() iterates activeControllers on handlerThread. Using
plain mutableMapOf() can throw ConcurrentModificationException in the field.

Replace all three maps with ConcurrentHashMap so both threads can access them
without locking while preserving correct cleanup semantics in onDestroy.

0xbrayo committed:
fix: drop unused SCHEDULE_EXACT_ALARM permission

Both alarm sites (SyncScheduler, UsageStatsWatcher) use setInexactRepeating,
so the exact-alarm permission was never exercised. It is a Play Console
sensitive permission that triggers a policy declaration requirement, so
requesting it needlessly risks release rejection.

0xbrayo committed:
fix: prevent overlapping-session double-count in SessionParser

Strict RESUME->PAUSE pair-matching scanned forward past other apps' events,
so a session absorbed time another app spent in the foreground when events
interleaved, inflating total screen time. Combined with the lastUpdated+1ms
incremental cursor, a long session could also be dropped permanently.

Replace it with a sequential foreground state machine: track a single open
session and close it at whichever comes first, the app's own PAUSE or a
different app's RESUME. The trailing still-open session is not emitted; it is
captured on a later run once its PAUSE (or next RESUME) arrives, so the cursor
cannot duplicate or drop it.

TimeToBuildBob committed:
fix: handle alarm errors and test JNI artifact restore

TimeToBuildBob commented on mobile/src/main/java/net/activitywatch/android/watcher/AlarmReceiver.kt:
```diff
@@ -12,16 +15,28 @@ class AlarmReceiver : BroadcastReceiver() {
+        } else if (intent.action == "net.activitywatch.android.watcher.LOG_DATA") {
             Log.w(TAG, "Action ${intent.action}, running sendHeartbeats")
-            if(UsageStatsWatcher.isUsageAllowed(context)) {
-                usw.sendHeartbeats()
+            if (UsageStatsWatcher.isUsageAllowed(context)) {
+                val pendingResult = goAsync()
+                CoroutineScope(Dispatchers.IO).launch {
+                    try {
+                        usw.sendHeartbeatsSuspend()
+                    } finally {
+                        pendingResult.finish()
+                    }
+                }
             }
         } else {
```
Fixed in `9030f6f`: `AlarmReceiver` now catches `Exception` around `sendHeartbeatsSuspend()` and still finishes the pending result, so server-startup JSON errors are logged instead of escaping the coroutine and crashing the app.

0xbrayo committed:
fix: restore SCHEDULE_EXACT_ALARM permission for widget updates

CategoryTimeWidgetProvider.schedulePeriodicUpdates() uses
setExactAndAllowWhileIdle(), which requires SCHEDULE_EXACT_ALARM on Android
12+. Without the permission it always hits the SecurityException fallback and
silently downgrades the widget's 5-minute refresh to inexact (Doze-delayed)
alarms. Restore the permission so exact scheduling works as intended.

0xbrayo committed:
refactor: dedupe bucket-id constants and guard EventParsingWorker on usage access

- Alias UsageStatsWatcher's bucket_id/unlock_bucket_id to SessionEventWatcher's
  SESSION_BUCKET_ID/UNLOCK_BUCKET_ID so the two watchers cannot drift to
  different bucket names.
- Skip EventParsingWorker cleanly (Result.success) when usage access is not
  granted, instead of constructing RustInterface and querying with no data.

0xbrayo committed:
fix: move MediaWatcher bucket init off main thread; recycle WebWatcher node

- MediaWatcher.onCreate() constructed RustInterface and called createBucketHelper()
  (blocking JNI/HTTP to a possibly-unstarted local server) on the
  NotificationListenerService main thread, risking an ANR. Post it to the
  handlerThread and mark ri @Volatile for safe cross-thread visibility.
- WebWatcher.onAccessibilityEvent used findWebView()'s returned node then dropped
  it without recycle(), leaking one AccessibilityNodeInfo per event on API 26-32.
  Recycle it unless it is the framework-owned event.source.

TimeToBuildBob committed:
fix(ci): stabilize aw-android web watcher e2e

TimeToBuildBob committed:
fix(ci): keep e2e runner script in one shell

TimeToBuildBob committed:
fix(ci): avoid multi-line shell construct in e2e runner script

The reactivecircus/android-emulator-runner@v2 action's parseScript()
splits the input by newlines and runs each as a separate sh -c command.
The original script wrapped commands in a bash -c with if/fi block that
fails when each line is a standalone invocation.

Replace the multi-line if/fi with single-line test && cmd || true
patterns. Remove the unnecessary bash -o pipefail wrapper — the >-
YAML folding already ensures the entire script runs as one sh -c call,
so  stays in scope for the final exit.

TimeToBuildBob committed:
fix(service): call startForeground() in onCreate() to avoid 5s OS timeout

On slow CI runners without KVM, onStartCommand() may not be invoked
within Android's 5-second startForegroundService() deadline, causing
RemoteServiceException and crashing the E2E ScreenshotTest.

Moving startForeground() to onCreate() promotes the service to foreground
immediately upon creation, before the countdown begins.

TimeToBuildBob committed:
Revert "fix(ci): avoid multi-line shell construct in e2e runner script"

This reverts commit 2a369d6c0d7471b25192845497209a360c5517a1.

TimeToBuildBob committed:
fix(ci): call startForeground() before native init to avoid 5s ANR

RustInterface() loads libaw_server.so + calls initialize() via JNI, which
can take several seconds on slow/no-KVM emulators (CI).  Android's
startForegroundService() timer started ticking before onCreate() ran,
so if native init took >5s the OS killed the process with
RemoteServiceException: startForegroundService without startForeground.

Fix: call startForeground() first in onCreate(), then do native init.
Also guard onDestroy() with isInitialized in case onCreate() throws.

TimeToBuildBob committed:
fix(media): handle bucket init failures

TimeToBuildBob commented on mobile/src/main/java/net/activitywatch/android/watcher/MediaWatcher.kt:
```diff
@@ -0,0 +1,302 @@
+            ri = r
+        }
+
+        val localRunnable = object : Runnable {
+            override fun run() {
+                pollActiveSessions()
+                handler?.postDelayed(this, POLL_INTERVAL_MS)
+            }
+        }
```
Fixed in `310747d`: the handler-thread RustInterface/media bucket initialization is now wrapped in `try/catch`, so a transient server-unavailable JSONException is logged instead of escaping the HandlerThread looper and crashing the app.

TimeToBuildBob committed:
fix(web): guard WebWatcher bucket initialization

TimeToBuildBob commented on mobile/src/main/java/net/activitywatch/android/watcher/WebWatcher.kt:
```diff
@@ -0,0 +1,174 @@
+    // TODO: This method is called very often, which might affect performance. Future optimizations needed.
+    override fun onAccessibilityEvent(event: AccessibilityEvent) {
+        if (shouldIgnoreEvent(event)) {
+            return
+        }
+
```
Fixed in `8747d8e`: `WebWatcher.onCreate()` now initializes `RustInterface` and `createBucketHelper()` on a `HandlerThread` inside `try/catch`, with `ri` published via `@Volatile` and the handler thread cleaned up in `onDestroy()`.

That keeps transient aw-server startup / non-JSON `JSONException` failures from escaping the `AccessibilityService` lifecycle and crashing the service. I also added the missing `super.onCreate()` call while touching this lifecycle method.

Verified: `git diff --check` passes. Local Gradle compile could not run in this environment because Java is not installed (`JAVA_HOME` unset and no `java` on `PATH`).

TimeToBuildBob committed:
fix(activity): skip main startup during onboarding

TimeToBuildBob commented on mobile/src/main/java/net/activitywatch/android/watcher/WebWatcher.kt:
```diff
@@ -0,0 +1,162 @@
+                        ?.let { handleWindowTitle(it.text.toString()) }
+                }
+            }
+        } catch(ex : Exception) {
+            Log.e(tag, ex.message!!)
+        }
```
Fixed in current code — line 118 now uses `ex.message ?: ex.toString()` so the catch block can't throw NPE.

TimeToBuildBob commented on mobile/src/main/java/net/activitywatch/android/RustInterface.kt:
```diff
@@ -82,25 +86,74 @@ class RustInterface constructor(context: Context? = null) {
+     * @param timestamp The exact start time of the event
+     * @param duration The precise duration in seconds
+     * @param data Event metadata (app name, package, etc.)
+     */
+    fun insertEvent(bucket_id: String, timestamp: Instant, duration: Double, data: JSONObject) {
+        val event = Event(timestamp, duration, data)
+        val msg = heartbeat(bucket_id, event.toString(), 0.0)
     }
```
Fixed — `insertEvent` now has `Log.d(TAG, "insertEvent response: $msg")` so native failures are logged.

TimeToBuildBob commented on mobile/src/main/java/net/activitywatch/android/MainActivity.kt:
```diff
@@ -56,15 +58,29 @@ class MainActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelecte
+        })
+
+        // Test RustInterface functions (remove after testing)
+        RustInterface(this).test()
```
Fixed — `RustInterface(this).test()` call is gone from `MainActivity.kt` in current code; no test buckets created on startup.
